docs/enhancements/scalable-syncset-processing.md

```diff
@@ -0,0 +1,91 @@
+
+Individual apply time appears to be approximately 0.2s - 1.6s, although 99% of all applys complete in under 0.5s. 0.5s * 600 = 300 seconds to process an entire cluster.
+
+Attempts to increase the goroutines to 100 lead to pod instability, at present problems with allocating OS threads.
+
+Moving to a design where we can horizontally scale the number of pods processing syncsets should let us scale out to meet the requirements of heavier SyncSet usage. In doing so we may be able to push far beyond the 1k ClusterDeployment limit and possibly save substantial cost for extremely large Hive deployments now requiring fewer shards.
+
+
+### Goals
+
+  1. Eliminate the SyncSet processing bottleneck in large Hive environments and enable horizontal scaling.
+
+### Non-Goals
+
+1. Moving SyncSets to a separate and independent project. This has been discussed for a long time and would overall be great. However this proposal is a step in this direction, the implementation would still need to be refined and extracted into a separate project to accomplish this fully.
+1. Use of a separate datastore. The data implications of processing SyncSets and storing results for the apply operations is still extremely questionable for etcd storage. In the future, it would be nice to have this data in another more suitable datastore. However this is not considered for this proposal and is being explicitly avoided.
+1. Opt-in to the new SyncSet method. This would be the only implementation moving forward, we will not maintain both the current and the new.
+
+## Open Questions
+
+RabbitMQ is not presently available on OperatorHub. [An operator](https://www.rabbitmq.com/kubernetes/operator/operator-overview.html) does however exist. How could Hive consume this operator and RabbitMQ safely?
+
+## Proposal
+
+Add a message bus to Hive deployments for processing SyncSets. Consumer pods will subscribe to the message bus queue and consume messages on the bus which are requests to apply resources.
+
+## Implementation Details / Notes / Constraints
+
+Deploy a RabbitMQ message bus along-side Hive.
+
+Hive's ClusterSync controller (or equivalent) will serve as the publisher of messages on the bus to a work queue. Each message will contain an identifier for the ClusterDeployment and possibly the SyncSet to process.
+
+Subscriber pods will process these messages, each message will be delivered to only one of the subscriber pods. (a supported configuration with RabbitMQ) Subscriber pods will initially leverage the normal use of Kubernetes clients and caching such that the actual data for the SyncSet resources and cluster kubeconfig Secrets are ready in memory when the message arrives. 
+
+Messages will initially be a request to sync *everything* for a cluster, exactly as the clustersync controller operates today. 
+* Doing this will allow us to have our subscriber pods write to the ClusterSync CR status as the controller does today, which means the API functions as before for users. 
```

Comment:
The piece that I see missing in this is something that ensures that only one subscriber is processing a given ClusterDeployment at once. There would need to be come coordination such that the clustersync controller knows that a request has already been sent for the ClusterDeployment and that a new one should not be sent until processing is complete on the first one.